File: src/features/typing/data/ts/linked_list_cycle_ii.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val = 0, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

function detectCycle(head: ListNode | null): ListNode | null {
    let slow = head;
    let fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow!.next;
        fast = fast.next.next;
        if (slow === fast) {
            let cur: ListNode | null = head;
            while (cur !== slow) {
                cur = cur!.next;
                slow = slow!.next;
            }
            return cur;
        }
    }
    return null;
}

File: src/features/typing/data/dart/linked_list_cycle_ii.dart
class Solution {
  ListNode? detectCycle(ListNode? head) {
    ListNode? slow = head;
    ListNode? fast = head;
    while (fast != null && fast.next != null) {
      slow = slow!.next;
      fast = fast.next!.next;
      if (slow == fast) {
        ListNode? cur = head;
        while (cur != slow) {
          cur = cur!.next;
          slow = slow!.next;
        }
        return cur;
      }
    }
    return null;
  }
}
